gemonlinedb/src/common/ConfigurationManager.cc

```diff
@@ -0,0 +1,167 @@
+#include "gem/onlinedb/ConfigurationManager.h"
+
+#include <cstdlib>
+
+#include <xercesc/dom/DOM.hpp>
+#include <xercesc/parsers/XercesDOMParser.hpp>
+
+#include "gem/onlinedb/ConfigurationLinker.h"
+#include "gem/onlinedb/SystemTopology.h"
+#include "gem/onlinedb/XMLConfigurationProvider.h"
+#include "gem/onlinedb/detail/FileUtils.h"
+
+namespace gem {
+    namespace onlinedb {
+
+        namespace /* anonymous */ {
+
+            /**
+             * \brief Loads a system-topology.xml file
+             */
+            DOMDocumentPtr loadXMLTopologyFile(const std::string &path)
+            {
+                using namespace detail::literals;
+                XERCES_CPP_NAMESPACE_USE
+
+                detail::XercesGuard guard; // Make sure that Xerces is loaded
+
+                // Find the schema (throws if not found)
+                auto schemaPath = detail::getFileInPath(
+                    "system-topology.xsd",
+                    "xml/schema:/opt/cmsgemos/share/gemonlinedb/xml/schema");
+
+                // Find the file (throws if not found)
+                auto filePath = detail::getFileInPath("system-topology.xml", path);
+
+                // Get an implementation
+                auto impl = DOMImplementationRegistry::getDOMImplementation("LS"_xml);
+                if (impl == nullptr) {
+                    throw std::logic_error("Cannot find a DOM implementation");
+                }
+
+                auto errorHandler =
+                    std::make_shared<detail::XercesAlwaysThrowErrorHandler>();
+
+                XercesDOMParser parser;
+                parser.setErrorHandler(errorHandler.get());
+                parser.setExternalSchemaLocation(
+                    ("https://github.com/cms-gem-daq-project/cmsgemos/tree/master/gemonlinedb/xml/schema/system-topology.xsd "
+                    + schemaPath).c_str());
+                parser.setValidationScheme(XercesDOMParser::Val_Always);
+                parser.setDoNamespaces(true);
+                parser.setDoSchema(true);
+
+                // Parse and validate
+                parser.parse(detail::XercesString(filePath));
+
+                // Load contents
+                DOMDocumentPtr document;
+                document.reset(parser.adoptDocument());
+                return document;
+            }
+
+        } // anonymous namespace
+
+        std::unique_ptr<ConfigurationManager>
+        ConfigurationManager::s_instance = nullptr;
+
+        boost::shared_mutex ConfigurationManager::s_mutex;
+
+        ConfigurationManager::ConfigurationManager(Source objectSource,
+                                                   Source topologySource)
+        {
+            try {
+                detail::xercesExceptionsToXcept([&]{
+                    // Deduce the search path for XML files
+                    std::string path = "/etc/opt/cmsgemos"; // FHS 3.0 standard location
```

Comment:
`/etc/opt/cmsgemos.d` is created by the RPM build process, with an optional `/etc/opt/cmsgemos.conf` as some sort of package configuration probably
I might prefer `/opt/cmsgemos/etc/...` as a location (everything can then be referenced from `CMSGEMOS_ROOT`), but in the end, if it's packaged correctly, it won't matter

Reply:
Referring to [section 3.7.4](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s07.html#etcoptConfigurationFilesForOpt) of the Filesystem Hierarchy Standard, the path should be `/etc/opt/...`. The reasoning behind the standard seems to be that the complete configuration from a given host resides in one directory that can be backed up independently, or copied over to a clean install.

Files in the config path are host-specific in nature (and optional: XML will not be used at P5) so the packages do not include them. The intent is that sysadmins will maintain a set of files matching their setup in `/etc/opt/cmsgemos`. When users need to override (some of) them, they can set the environment variable with the same syntax as a traditional `PATH`.

`.d` directories usually have a special meaning: configuration fragments found in these directories will be merged together to produce the final config. `gemonlinedb` doesn't support this method of merging. Since `/etc/opt/cmsgemos.d` appears not to be used at all, I would  replace it by `/etc/opt/cmsgemos`. Most directories in `/etc` do not use the `.d` suffix anyway.

Reply:
The close linking between the build process for `cmsgemos` and `xdaq` would encourage adoption of a similar structure
```
[]% ll /opt/xdaq/
total 40
drwxr-xr-x.  2 root root  4096 Feb  1  2018 bin
drwxr-xr-x.  2 root root  4096 Feb  1  2018 config
drwxr-xr-x.  3 root root  4096 Feb  1  2018 etc
drwxr-xr-x. 50 root root  4096 Feb 14 15:13 htdocs
drwxr-xr-x. 59 root root  4096 Feb  1  2018 include
drwxr-xr-x.  4 root root 12288 Feb 22 12:15 lib
drwxr-xr-x.  2 root root  4096 Dec 16  2016 sbin
drwxr-xr-x.  4 root root  4096 Oct 25  2017 share
[]% ll /opt/xdaq/etc
total 48
-rw-r-xr-x. 1 root root  1826 Nov 14  2017 base.profile
drwxr-xr-x. 5 root root  4096 Dec 19  2016 CAENUpgrader
-rw-r-xr-x. 1 root root  3909 Nov 14  2017 default.profile
-rw-r-xr-x. 1 root root  1899 Nov 14  2017 default.profile.macosx
-rw-r-xr-x. 1 root root   462 Dec 16  2016 htaccess
-rw-r-xr-x. 1 root root    53 Dec 16  2016 htgroup
-rw-r-xr-x. 1 root root    19 Dec 16  2016 htpasswd
-rw-r-xr-x. 1 root root   255 Dec 16  2016 log4cplus.properties
-rwxr-xr-x. 1 root root 10765 Dec 16  2016 tinyproxy.conf
-rw-r-xr-x. 1 root root   435 Dec 16  2016 udpServer.sh
[]% ll /opt/cactus
total 16
drwxr-xr-x. 4 root root 4096 Aug  1  2018 bin
drwxr-xr-x. 4 root root 4096 Aug  1  2018 etc
drwxr-xr-x. 4 root root 4096 Aug  1  2018 include
drwxr-xr-x. 3 root root 4096 Aug  3  2018 lib
[]% ll /opt/cactus/etc
total 8
drwxr-xr-x. 2 root root 4096 Jan 31  2018 amc13
drwxr-xr-x. 4 root root 4096 Aug  1  2018 uhal
[]% ll /etc/opt
total 8
drwxr-x---. 2 gempro gempro 4096 Feb 14 15:55 cmsgemos.d
drwxr-xr-x. 9 root   root   4096 Mar 24  2018 rh
```

Though, on the other hand, the `xdaq` zone configurations are installed in
```
/etc/xdaq.d
```

So, given that
```
/opt/cmsgemos/etc
```
and 
```
/etc/opt/cmsgemos.d
```
are already created (and at least one of them used), unless there is a reasonable concern that you're attempting to resolve, one of the two should be used.